cmd/postgres_exporter/postgres_exporter.go

```diff
@@ -1425,8 +1425,84 @@ func main() {
 		exporter.servers.Close()
 	}()
 
-	prometheus.MustRegister(exporter)
+	exporter_shared.RunServer("PostgreSQL", *listenAddress, *metricPath, newHandler(map[string]prometheus.Collector{
+		"exporter": exporter,
+	}))
+}
+
+// handler wraps an unfiltered http.Handler but uses a filtered handler,
+// created on the fly, if filtering is requested. Create instances with
+// newHandler. It used for collectors filtering.
+type handler struct {
+	unfilteredHandler http.Handler
+	collectors        map[string]prometheus.Collector
+}
+
+func newHandler(collectors map[string]prometheus.Collector) *handler {
+	h := &handler{collectors: collectors}
+
+	innerHandler, err := h.innerHandler()
+	if err != nil {
+		log.Fatalf("Couldn't create metrics handler: %s", err)
+	}
+
+	h.unfilteredHandler = innerHandler
+	return h
+}
+
+// ServeHTTP implements http.Handler.
+func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	filters := r.URL.Query()["collect[]"]
+	log.Debugln("collect query:", filters)
+
+	if len(filters) == 0 {
+		// No filters, use the prepared unfiltered handler.
+		h.unfilteredHandler.ServeHTTP(w, r)
+		return
+	}
+
+	filteredHandler, err := h.innerHandler(filters...)
+	if err != nil {
+		log.Warnln("Couldn't create filtered metrics handler:", err)
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte(fmt.Sprintf("Couldn't create filtered metrics handler: %s", err)))
```

Comment:
```go
fmt.Fprintf(w, "Couldn't create filtered metrics handler: %s", err)
```

Reply:
JFYI

Reply:
Thanks. I'll take note. I've ported it from `node_exporter` and I also not so pedantic to details, so it's the reason why it's made so.